zerver/views/streams.py

```diff
@@ -319,8 +319,10 @@ def add_subscriptions_backend(
         if user_profile.realm.is_zephyr_mirror_realm and not all(stream.invite_only for stream in streams):
             return json_error(_("You can only invite other Zephyr mirroring users to invite-only streams."))
         subscribers = set(principal_to_user_profile(user_profile, principal) for principal in principals)
-    else:
+    elif user_profile.realm.is_zephyr_mirror_realm:
         subscribers = set([user_profile])
+    else:
+        return json_error(_("Stream must have at least one subscriber."))
```

Comment:
I think this is making `./tools/test-api` fail. Could you please investigate? :)

Reply:
Okay, so I think there is some confusion here:
* I think what our goal here is to modify who gets subscribed to a stream at the time of the creation of the stream, for which this is the wrong function to change. `add_subscriptions_backend` is a general function for adding subscriptions, it has nothing to do with stream creation.
* The function that should raise this error is the one where a stream is created. You need to control the subscribers at the time of stream creation, not at any time a subscription is added (which is what `add_subscriptions_backend` is for).

Therefore, I would recommend moving these changes to the file where the stream is created. For instance, when a new stream is created, which function calls this one? That's where you would check whether `principals` is 0, and if it is, raise this error! :)

Does that make sense?

Reply:
@eeshangarg I get your point. But from my debugging I think `add_subscriptions_backend` is the function which is get called on stream creation.
As you can see in `stream_create.js` file when user click on submit button the data is posted to url `/json/users/me/subscriptions`.
```javascript
function ajaxSubscribeForCreation(stream_name, description, principals, invite_only, announce) {
    // Subscribe yourself and possible other people to a new stream.
    return channel.post({
        url: "/json/users/me/subscriptions",
        data: {subscriptions: JSON.stringify([{name: stream_name,
                                               description: description}]),
               principals: JSON.stringify(principals),
               invite_only: JSON.stringify(invite_only),
               announce: JSON.stringify(announce),
        },
        success: function () {
            $(".stream_change_property_info").hide();
            $("#create_stream_name").val("");
            $("#create_stream_description").val("");
            loading.destroy_indicator($('#stream_creating_indicator'));
            // The rest of the work is done via the subscribe event we will get
        },
```

The function for POST request to this url is 
```python
    url(r'^users/me/subscriptions$', rest_dispatch,
        {'GET': 'zerver.views.streams.list_subscriptions_backend',
         'POST': 'zerver.views.streams.add_subscriptions_backend',
         'PATCH': 'zerver.views.streams.update_subscriptions_backend',
         'DELETE': 'zerver.views.streams.remove_subscriptions_backend'}),
```
Let me know if I am getting this wrong.

Reply:
@YJDave: Okay, so yes exactly. This error needs to be raised in `ajaxSubscribeForCreation` if `principals` is 0. This confirms that you shouldn't need to touch `add_subscriptions_backend`. :)

Reply:
@eeshangarg that error is only for frontend, but we should also have the same behavior in backend. Other than this, one reason to not allowing zero subscribers backend is that
* we can't subscribe current user.
* And our code in `actions.py` function `bulk_add_subscriptions` which is called in `add_subscriptions_backend` does suppose that principal length is more than zero and it do breaks if it is not.
You can checkout out this [message](https://chat.zulip.org/#narrow/stream/backend/subject/stream.20subscription/near/452684)

Reply:
@YJDave: I get what you mean. But like I said, that backend function is a generic function for subscriptions and is also a part of our REST API. What you are trying to accomplish is to restrict who gets subscribed when a stream is created, not when anytime a subscription is added.

Technically, you don't even need to touch the backend for this change. There is no Python backend that controls who gets subscribed when a stream is created except `ajaxSubscribeForCreation`. Does that make it more clear?

Reply:
To repeat, you only need to change `ajaxSubscribeForCreation` to complain when `principals` is zero. This change has nothing to do with how subscriptions work in general, it is about how they should work when a stream is created.

Reply:
Okay, got it. Will update soon!